### PR TITLE
DNS support [13406]

### PR DESCRIPTION
**Pull request by jparisu**
Signed-off-by: jparisu <javierparis@eprosima.com>

### CODE REVIEW

File: include/ddsrouter/types/address/Address.hpp [hunk 1]
```diff
@@ -39,6 +39,7 @@ using PortType = uint16_t;
 //! Different versions allowed for IP
 enum IpVersion
 {
+    NONE_IPv,
```
Comment (juanlofer-eprosima) [inline include/ddsrouter/types/address/Address.hpp:42]: I do not think this new category is really necessary.

File: src/cpp/configuration/implementations/Address_configuration.cpp [hunk 2]
```diff
@@ -30,12 +30,83 @@ Address::Address(
...
+        catch (const std::exception& e)
+        {
+            throw ConfigurationException(utils::Formatter() <<
+                          "Error getting Address ip version: " << e.what());
+        }
+    }
+    else
+    {
+        // Set ip version default
+        ip_version_ = NONE_IPv;
```
Comment (juanlofer-eprosima) [inline src/cpp/configuration/implementations/Address_configuration.cpp:68]:
```suggestion
        ip_version_ = Address::default_ip_version();
```

File: src/cpp/configuration/implementations/Address_configuration.cpp [hunk 3]
```diff
@@ -30,12 +30,83 @@ Address::Address(
...
+                    {
+                        ip_ = response.second.begin()->data();
+                    }
+                    else
+                    {
+                        throw ConfigurationException(utils::Formatter() <<
+                          "Incorrect IPv6 set or DNS not found: " << ip_);
+                    }
+                }
+                else
```
Comment (juanlofer-eprosima) [inline src/cpp/configuration/implementations/Address_configuration.cpp:97]: Here both ``NONE_IPv`` and ``IPv4`` behave the same, that is why I don't think the new category is useful.

File: src/cpp/configuration/implementations/Address_configuration.cpp [hunk 4]
```diff
@@ -48,6 +119,24 @@ Address::Address(
         ip_ = Address::default_ip();
```
Comment (juanlofer-eprosima) [inline src/cpp/configuration/implementations/Address_configuration.cpp:119]: Add ip_version as argument for this function, and define a default static ipv6 address.

File: src/cpp/configuration/implementations/Address_configuration.cpp [hunk 5]
```diff
@@ -48,6 +119,24 @@ Address::Address(
...
     }
 
+    // Check the IP Version is correctly set
```
Comment (juanlofer-eprosima) [inline src/cpp/configuration/implementations/Address_configuration.cpp:122]: Remove so checks are also performed for the case when ip_version is provided.

File: src/cpp/configuration/implementations/Address_configuration.cpp [hunk 6]
```diff
@@ -48,6 +119,24 @@ Address::Address(
...
+    if (ip_version_ == NONE_IPv)
+    {
+        if (Address::is_ipv4_correct(ip_))
+        {
+            ip_version_ = IPv4;
+        }
+        else if (Address::is_ipv6_correct(ip_))
+        {
+            ip_version_ = IPv6;
```
Comment (juanlofer-eprosima) [inline src/cpp/configuration/implementations/Address_configuration.cpp:131]: Is this reachable?

**Review by juanlofer-eprosima — CHANGES_REQUESTED** [review]
Please add/modify documentation.

File: src/cpp/configuration/implementations/Address_configuration.cpp (hunk 5 above)
Comment (juanlofer-eprosima) [reply src/cpp/configuration/implementations/Address_configuration.cpp:122]: Although just removing the conditional would be not enough, as ip_version could change from the one provided without the user noticing.

File: include/ddsrouter/types/address/Address.hpp (hunk 1 above)
Comment (jparisu) [reply include/ddsrouter/types/address/Address.hpp:42]: :+1:

File: src/cpp/configuration/implementations/Address_configuration.cpp (hunk 2 above)
Comment (jparisu) [reply src/cpp/configuration/implementations/Address_configuration.cpp:68]: old version

File: src/cpp/configuration/implementations/Address_configuration.cpp (hunk 3 above)
Comment (jparisu) [reply src/cpp/configuration/implementations/Address_configuration.cpp:97]: :+1:

File: src/cpp/configuration/implementations/Address_configuration.cpp (hunk 4 above)
Comment (jparisu) [reply src/cpp/configuration/implementations/Address_configuration.cpp:119]: :+1:

File: src/cpp/configuration/implementations/Address_configuration.cpp (hunk 5 above)
Comment (jparisu) [reply src/cpp/configuration/implementations/Address_configuration.cpp:122]: old version

File: src/cpp/configuration/implementations/Address_configuration.cpp (hunk 6 above)
Comment (jparisu) [reply src/cpp/configuration/implementations/Address_configuration.cpp:131]: I have dealt with it

File: test/unittest/configuration/AddressConfigurationTest.cpp [hunk 7]
```diff
@@ -0,0 +1,324 @@
...
+    }
+
+    // Specific DNS addresses IPv6
+    // This test could fail, as not every local DNS recognize ::1 as localhost IPv6
+    // {
+    //     IpType dns_name = "localhost";
+    //     IpType expected_address = "::1";
+
+    //     RawConfiguration yaml;
+    //     yaml[ADDRESS_IP_TAG] = dns_name;
```
Comment (juanlofer-eprosima) [inline test/unittest/configuration/AddressConfigurationTest.cpp:255]:
```suggestion
    //     yaml[ADDRESS_DNS_TAG] = dns_name;
```

File: test/unittest/configuration/AddressConfigurationTest.cpp [hunk 8]
```diff
@@ -0,0 +1,324 @@
...
+        array_config.push_back("rand_val_1");
+        array_config.push_back("rand_val_2");
+
+        ASSERT_THROW(new Address(array_config), ConfigurationException);
+    }
+
+    // DNS fail to find network
+    {
+        RawConfiguration yaml;
+        yaml[ADDRESS_IP_TAG] = "thishostdoesnotexist_andmustnotexist.zzz";
```
Comment (juanlofer-eprosima) [inline test/unittest/configuration/AddressConfigurationTest.cpp:296]:
```suggestion
        yaml[ADDRESS_DNS_TAG] = "thishostdoesnotexist_andmustnotexist.zzz";
```

**Review by juanlofer-eprosima — CHANGES_REQUESTED** [review]

File: docs/rst/user_manual/configuration.rst [hunk 9]
```diff
@@ -150,6 +150,7 @@ An Address is defined by:
   If it is not set, it would be chosen by default depending on the Participant Type.
 * *IP version*: ``v4`` or ``v6``.
   If it is not set, it would be chosen depending on the *IP* string format.
+* *Domain Name*: Domain unique name to ask DNS server for the related IP.
```
[the remark this answers is not in the document]
Comment (jparisu) [reply docs/rst/user_manual/configuration.rst:153]: :+1:

File: docs/rst/user_manual/configuration.rst [hunk 10]
```diff
@@ -163,9 +164,10 @@ An Address is defined by:
     ip: "2001:4860:4860::8844"      # Recognized as IPv6
     port: 1616
 
-.. warning::
+    ################
 
-    ``ip`` field does not currently support DNS names, only well-formed IP addresses.
+    domain: "localhost"             # DNS call get: 127.0.0.1
```
[the remark this answers is not in the document]
Comment (jparisu) [reply docs/rst/user_manual/configuration.rst:169]: :+1:

File: include/ddsrouter/exceptions/ValueNotAllowedException.hpp [hunk 11]
```diff
@@ -0,0 +1,42 @@
+// Copyright 2021 Proyectos y Sistemas de Mantenimiento SL (eProsima).
```
[the remark this answers is not in the document]
Comment (jparisu) [reply include/ddsrouter/exceptions/ValueNotAllowedException.hpp:1]: :+1:

File: include/ddsrouter/types/utils.hpp [hunk 12]
```diff
@@ -194,6 +194,13 @@ bool are_set_of_ptr_equal(
     return true;
 }
 
+/**
+ * @brief Shit, This Should Not Happen.
+ *
+ * @param formatter msg of the unexpected case.
+ */
+void stsnh(const Formatter& formatter);
```
[the remark this answers is not in the document]
Comment (jparisu) [reply include/ddsrouter/types/utils.hpp:202]: :+1: 
Suggest me a name if you find a better one.

File: test/unittest/configuration/AddressConfigurationTest.cpp [hunk 13]
```diff
@@ -0,0 +1,324 @@
+// Copyright 2021 Proyectos y Sistemas de Mantenimiento SL (eProsima).
```
[the remark this answers is not in the document]
Comment (jparisu) [reply test/unittest/configuration/AddressConfigurationTest.cpp:1]: :+1:

File: src/cpp/configuration/implementations/Address_configuration.cpp [hunk 14]
```diff
@@ -45,9 +139,106 @@ Address::Address(
...
+        // If IP was specified, do not get domain
+        if (ip_has_been_set)
+        {
+            logWarning(
+                DDSROUTER_ADDRESS_CONFIGURATION,
+                "Ip was specified, domain tag skipped.");
+        }
+        else
+        {
+            std::string domain_name = configuration[ADDRESS_DNS_TAG].as<std::string>();
```
[the remark this answers is not in the document]
Comment (jparisu) [reply src/cpp/configuration/implementations/Address_configuration.cpp:161]: Good catch :+1:

File: src/cpp/configuration/implementations/Address_configuration.cpp [hunk 15]
```diff
@@ -45,9 +139,106 @@ Address::Address(
...
+                        "DNS address not found: " << domain_name);
+                }
+            }
+        }
+    }
+
+    ////////////////
+    // GETTING DEFAULT IP
+
+    // Check the IP Version is correctly set
```
[the remark this answers is not in the document]
Comment (jparisu) [reply src/cpp/configuration/implementations/Address_configuration.cpp:229]: :+1:

File: test/unittest/configuration/AddressConfigurationTest.cpp (hunk 7 above)
Comment (jparisu) [reply test/unittest/configuration/AddressConfigurationTest.cpp:255]: :+1:

File: test/unittest/configuration/AddressConfigurationTest.cpp (hunk 8 above)
Comment (jparisu) [reply test/unittest/configuration/AddressConfigurationTest.cpp:296]: :+1:

File: include/ddsrouter/types/utils.hpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (juanlofer-eprosima) [reply include/ddsrouter/types/utils.hpp]: I like the new one.

File: src/cpp/configuration/implementations/Address_configuration.cpp (hunk 14 above)
[the remark this answers is not in the document]
Comment (juanlofer-eprosima) [reply src/cpp/configuration/implementations/Address_configuration.cpp]: Well tried :)

File: include/ddsrouter/types/utils.hpp [hunk 16]
```diff
@@ -194,6 +194,20 @@ bool are_set_of_ptr_equal(
     return true;
 }
 
+/**
+ * @brief This Should Not Happen.
+ *
+ * This method should be called when something that should not have happened at all happen.
```
Comment (juanlofer-eprosima) [inline include/ddsrouter/types/utils.hpp:200]:
```suggestion
 * This method should be called when something that should not have happened at all happens.
```

File: include/ddsrouter/types/utils.hpp [hunk 17]
```diff
@@ -194,6 +194,20 @@ bool are_set_of_ptr_equal(
...
+ * This will show an error log, assert false and throw an exception.
+ *
+ * Do not use this method when the error could come from user or output interaction, it should only be used
+ * for inconsistency inside the program or C++ weird behaviours (enumeration values out of their range).
```
Comment (juanlofer-eprosima) [inline include/ddsrouter/types/utils.hpp:204]:
```suggestion
 * for inconsistency inside the program or C++ weird behaviours (e.g. enumeration values out of their range).
```

File: src/cpp/types/utils.cpp [hunk 18]
```diff
@@ -60,6 +63,14 @@ std::string Formatter::to_string() const noexcept
     return ss_.str().c_str();
 }
 
+void tsnh(
+        const Formatter& formatter)
+{
+    logError(DDSROUTER_TSNH, "This Should Have Not Happened: " << formatter.to_string());
```
Comment (juanlofer-eprosima) [inline src/cpp/types/utils.cpp:69]:
```suggestion
    logError(DDSROUTER_TSNH, "This Should Not Have Happened: " << formatter.to_string());
```

File: src/cpp/configuration/implementations/Address_configuration.cpp [hunk 19]
```diff
@@ -45,9 +139,115 @@ Address::Address(
...
     {
-        ip_ = Address::default_ip();
+        ip_has_been_set = false;
     }
 
+
+    ////////////////
+    // DNS
+
+    // Get IP
```
Comment (juanlofer-eprosima) [inline src/cpp/configuration/implementations/Address_configuration.cpp:149]: Change comment

**Review by juanlofer-eprosima — CHANGES_REQUESTED** [review]
NIT

File: src/cpp/configuration/implementations/Address_configuration.cpp (hunk 19 above)
Comment (jparisu) [reply src/cpp/configuration/implementations/Address_configuration.cpp:149]: :+1:

**Review by juanlofer-eprosima — DISMISSED** [review]
LGTM

**Review by juanlofer-eprosima — APPROVED** [review]
LGTM

**Review by juanlofer-eprosima — CHANGES_REQUESTED** [review]
Fix Windows tests

**Review by juanlofer-eprosima — APPROVED** [review]
LGTM

**Pull request merged**